apps/x509.c

```diff
@@ -571,6 +611,25 @@ int x509_main(int argc, char **argv)
             BIO_printf(bio_err, "error unpacking public key\n");
             goto end;
         }
+#ifndef OPENSSL_NO_SM2
+        if (sm2_id != NULL) {
+            ASN1_OCTET_STRING *v;
+
+            v = ASN1_OCTET_STRING_new();
+            if (v == NULL) {
+                BIO_printf(bio_err, "error: SM2 ID allocation failed\n");
+                goto end;
+            }
+
+            if (!ASN1_OCTET_STRING_set(v, sm2_id, sm2_idlen)) {
+                BIO_printf(bio_err, "error: setting SM2 ID failed\n");
+                ASN1_OCTET_STRING_free(v);
+                goto end;
+            }
+
+            X509_REQ_set0_sm2_id(req, v);
+        }
+#endif
         i = X509_REQ_verify(req, pkey);
```

Comment:
So it seems that what's *really* missing is a `do_X509_REQ_verify` that takes a sigopts stack...

Reply:
Yes, it seems that problem is how to set an `sm2-id` to a `X509_REQ` with `sig-opt`

Reply:
But that loses the ability of setting different IDs for CSR and final signing separately.

Reply:
@william-zk, it seems to go deeper, there's no X509_REQ_verify_ctx() either, so support seems to be lacking quite deeply.  I'm preparing a PR that adds such support.

Reply:
@InfoHunter, sorry, you lost me there...  are you essentially asking for vfyopts?  That would certainly be possible...

Reply:
@levitte , so what you meant was adding 'vfyopts' and then use its content in something like `do_X509_REQ_verify` ?